File: src/ios/core/APIs/Events/EventTarget.swift
//
//  EventTarget.swift
//  WTVJavaScriptCore
//
//  Created by Jing KE on 17/06/16.
//  Copyright © 2016 WizTiVi. All rights reserved.
//

import Foundation
import JavaScriptCore

@objc protocol EventTargetJSExport: JSExport {
    
    func addEventListener(type: String, _ callback: JSValue, _ options: JSValue?)
    func removeEventListener(type: String?, _ callback: JSValue?, _ options: JSValue?)
    func dispatchEvent(event: Event) -> Bool
    
}

class EventTarget: JSClassDelegate {
    
    override init(context: ScriptContext) {
        super.init(context: context)
        self.addEventListener("load", listener: EventListener.create(withHandler: self.onLoad))
        self.dispatchLoader()
    }
    
}

extension EventTarget: EventTargetJSExport {
    
    func addEventListener(type: String, _ callback: JSValue, _ options: JSValue? = nil) {
        self.addEventListener(type, listener: EventListener.create(callback))
    }
    
    func removeEventListener(type: String? = nil, _ callback: JSValue? = nil, _ options: JSValue? = nil) {
        if let type = type {
            if let callback = callback {
                var listeners = self.listeners[type] ?? Array<EventListener>()
                let callbacks = listeners.map({ return $0.callback ?? JSValue() })
                if let index = callbacks.indexOf(callback) {
                    let listener = listeners[index]
                    self.removeEventListener(type, listener: listener)
                }
            } else {
                self.removeEventListener(type, listener: nil)
            }
        } else {
            self.removeEventListener(nil, listener: nil)
        }
    }
    
    override func dispatchEvent(event: Event) -> Bool {
        event.attatchEventTarget(self)
        return super.dispatchEvent(event)
    }

}

extension EventTarget {
    
    func dispatchLoader() {
        let event = Event.create(self.context, type: "load", initDict: [:])
        self.dispatchEvent(event)
    }
    
}

extension EventTarget {
    
    func onLoad(event: Event) {
        guard let jsEvent = event.thisJSValue else { return }
        guard let this = self.thisJSValue else { return }
        guard let onload = this.objectForKeyedSubscript("onload") where !onload.isUndefined && !onload.isNull else { return }
        onload.callWithArguments([jsEvent])
    }
    
}

extension EventTarget {
    
    class func create(context: ScriptContext) -> EventTarget {
        return EventTarget(context: context)
    }
    
}

File: www/JSCore/Events/EventTarget.js
importScripts("JSClassDelegate");
/**
 * @interface	{EventTarget}
 */
class EventTarget extends JSClassDelegate {

    constructor(instance) {
        super();
        if (this.constructor.name === "EventTarget") {
            throw new Error(`Illegal constructor - ${this.constructor.name}`);
        }
    }

    get onload() { return this.__onload__; }
    set onload(handler) { this.__onload__ = handler; this.instance.registerThisJSValue(this); }

    /**
     * @parameter	{String} type
     * @parameter	{EventListener} callback
     * @parameter	{AddEventListenerOptions|Boolean} options
     */
    addEventListener(type, callback, options) {
        this.instance && this.instance.addEventListener(type, callback, options);
        return this;
    }

    /**
     * @parameter	{String} type
     * @parameter	{EventListener} callback
     * @parameter	{AddEventListenerOptions|Boolean} options
     */
    removeEventListener(type, callback, options) {
        this.instance && this.instance.removeEventListener(type, callback, options);
        return this;
    }

    /**
     * @parameter	{Event} event
     * @returns     {Boolean}  
     */
    dispatchEvent(event) {
        if (!this.instance) { return false; }
        if (typeof event === "string") { 
            event = this.instance.getEventByType(event); 
            if (event) {
                return this.instance.dispatchEvent(event);
            }
        }
        if (!event) { return false; }
        return this.instance.dispatchEvent(event.instance);
    }

    // registerEvent(event) {
    //     this.instance && this.instance.registerEvent(event.instance);
    // }

    // getEventByType(type) {
    //     return this.instance &&  this.instance.getEventByType(type).thisJSValue;
    // }

    __onload__(event) { console.info(`[${this.constructor.name}] loaded`, event); }

}